File: 169-majority-element/majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int n = nums.size();
        int flag;
        if(n % 2 == 0)
            flag = n / 2;
        else
            flag = (n/2) + 1;
        if(nums.size() <= 2)
            return nums[0];
        sort(nums.begin(), nums.end());
        int count = 1;
        for(auto i=1; i<n; i++){
            if(nums[i] == nums[i-1])
                count++;
            else 
                count = 1;
            if(count >= flag)
                return nums[i];
        }
        return 0;
    }
};

File: 169-majority-element/majority-element.java
//use two variables candidate = nums[0] and count = 1 and start from 2nd element if count == 0 new candidate , if element == candidate +1, else -1
class Solution {
    public int majorityElement(int[] nums) {
        int n = nums.length;
        int candidate = nums[0], count = 1;
        for(int i=1; i<n; i++){
            if(count > n/2)
                return candidate;
            if(count == 0){
                candidate = nums[i];
                count = 1;
            }else if(candidate == nums[i]){
                count++;
            }else{
                count--;
            }
        }
        return candidate;
    }
}